azurelinuxagent/common/logcollector.py

```diff
@@ -0,0 +1,302 @@
+            if ext in [".gz", ".zip", ".xz"]:
+                LogCollector._log_to_results_file("Discarding large binary file {0}".format(file_path))
+                return None
+
+            truncated_file_path = os.path.join(_TRUNCATED_FILES_DIR, file_path.replace(os.path.sep, "_"))
+            if os.path.exists(truncated_file_path):
+                original_file_mtime = os.path.getmtime(file_path)
+                truncated_file_mtime = os.path.getmtime(truncated_file_path)
+
+                # If the original file hasn't been updated since the truncated file, it means there were no changes
+                # and we don't need to truncate it again.
+                if original_file_mtime < truncated_file_mtime:
+                    return truncated_file_path
+
+            # Get the last N bytes of the file
+            with open(truncated_file_path, "w+") as fh:
+                LogCollector.run_shell_command(["tail", "-c", str(_FILE_SIZE_LIMIT), file_path], stdout=fh)
+
+            return truncated_file_path
+        except OSError as e:
+            LogCollector._log_to_results_file("Failed to truncate large file: {0}".format(ustr(e)))
+            return None
+
+    def _get_file_priority(self, file):
+        # The sooner the file appears in the must collect list, the bigger its priority.
+        # Priority is higher the lower the number (0 is highest priority).
+        if file in self.must_collect_files:
+            return self.must_collect_files.index(file)
+        else:
+            # Doesn't matter, file is not in the must collect list, assign a low priority
+            return 999999999
+
+    def _get_priority_files_list(self, file_list):
+        # Given a list of files to collect, determine if they show up in the must collect list and build a priority
+        # queue. The queue will determine the order in which the files are collected, highest priority files first.
+        priority_file_queue = []
+        for file in file_list:
+            priority = self._get_file_priority(file)
+            heappush(priority_file_queue, (priority, file))
+
+        return priority_file_queue
+
+    def _get_final_list_for_archive(self, priority_file_queue):
+        # Given a priority queue of files to collect, add one by one while the archive size is under the size limit.
+        # If a single file is over the file size limit, truncate it before adding it to the archive.
+        self._log_to_results_file("\n### Preparing list of files to add to archive ###")
+        total_uncompressed_size = 0
+        final_files_to_collect = []
+
+        while priority_file_queue:
+            file_path = heappop(priority_file_queue)[1]  # (priority, file_path)
+            file_size = min(os.path.getsize(file_path), _FILE_SIZE_LIMIT)
+
+            if total_uncompressed_size + file_size > _UNCOMPRESSED_ARCHIVE_SIZE_LIMIT:
+                self._log_to_results_file("Archive too big, done with adding files.")
+                break
```

Comment:
you need to truncate first, then do the size check, right?

Reply:
Already handled in L236, `file_size = min(os.path.getsize(file_path), _FILE_SIZE_LIMIT)`.